pkg/awsvpc/sg_analysis.go

```diff
@@ -0,0 +1,228 @@
+	connStr := fmt.Sprintf("protocol: %s,  icmpType: %s", *ruleObj.IpProtocol, conns)
+	remote, err := sga.getRemoteCidr(ruleObj.IpRanges, ruleObj.UserIdGroupPairs)
+	if err != nil {
+		return "", nil, err
+	}
+	ruleRes = &commonvpc.SGRule{
+		Connections: conns,
+		Remote:      &commonvpc.RuleTarget{Cidr: remote, SgName: ""},
+	}
+	ruleStr = getRuleStr(direction, connStr, ruleRes.Remote.Cidr.String())
+	return
+}
+
+func (sga *SpecificAnalyzer) GetSGRule(index int) (
+	ruleStr string, ruleRes *commonvpc.SGRule, isIngress bool, err error) {
+	var ruleObj types.IpPermission
+	direction := commonvpc.Inbound
+	if index < len(sga.sgResource.IpPermissions) {
+		isIngress = true
+		ruleObj = sga.sgResource.IpPermissions[index]
+	} else {
+		direction = commonvpc.Outbound
+		isIngress = false
+		ruleObj = sga.sgResource.IpPermissionsEgress[index-len(sga.sgResource.IpPermissions)]
+	}
+	switch *ruleObj.IpProtocol {
+	case commonvpc.AllProtocols: // all protocols
+		ruleStr, ruleRes, err = sga.getProtocolAllRule(&ruleObj, direction)
+	case commonvpc.ProtocolTCP:
+		ruleStr, ruleRes, err = sga.getProtocolTcpudpRule(&ruleObj, direction)
+	case commonvpc.ProtocolUDP:
+		ruleStr, ruleRes, err = sga.getProtocolTcpudpRule(&ruleObj, direction)
+	case commonvpc.ProtocolICMP:
+		ruleStr, ruleRes, err = sga.getProtocolIcmpRule(&ruleObj, direction)
+	default:
+		return "", nil, false, fmt.Errorf("getSGRule error: unsupported type")
+	}
+	if err != nil {
+		return "", nil, false, err
+	}
+	ruleRes.Local = ipblock.GetCidrAll()
+	ruleRes.Index = index
+	return fmt.Sprintf("index: %d, %v", index, ruleStr), ruleRes, isIngress, nil
+}
+
+func (sga *SpecificAnalyzer) GetSGrules() (ingressRules, egressRules []*commonvpc.SGRule, err error) {
+	ingressRules = []*commonvpc.SGRule{}
+	egressRules = []*commonvpc.SGRule{}
+	numRules := len(sga.sgResource.IpPermissions) + len(sga.sgResource.IpPermissionsEgress)
+	for index := 0; index < numRules; index++ {
+		_, ruleObj, isIngress, err := sga.GetSGRule(index)
+		if err != nil {
+			return nil, nil, err
+		}
+		if ruleObj == nil {
+			continue
+		}
+		if isIngress {
+			ingressRules = append(ingressRules, ruleObj)
+		} else {
+			egressRules = append(egressRules, ruleObj)
+		}
+	}
+	return ingressRules, egressRules, nil
+}
+
+// StringRules returns a string with the details of the specified rules
+func (sga *SpecificAnalyzer) StringRules(rules []int) string {
+	strRulesSlice := make([]string, len(rules))
+	for i, ruleIndex := range rules {
+		strRule, _, _, err := sga.GetSGRule(ruleIndex)
+		if err != nil {
+			return ""
+		}
+		strRulesSlice[i] = "\t" + strRule
+	}
+	sort.Strings(strRulesSlice)
+	return strings.Join(strRulesSlice, "")
+}
+
+func (sga *SpecificAnalyzer) SetSGmap(sgMap map[string]*commonvpc.SecurityGroup) {
+	sga.sgMap = sgMap
+}
+
+func (sga *SpecificAnalyzer) ReferencedIPblocks() []*ipblock.IPBlock {
+	return sga.referencedIPblocks
+}
+
+func (sga *SpecificAnalyzer) GetNumberOfRules() int {
```

Comment:
Why is this needed?

Reply:
because we print the number of rules in verbose mode, and we do that in commonvpc pkg where we do not have access to the rule objects.

Reply:
What about using  `GetRules()` or do you just need the number of rules per sa without printing other rules details?

Reply:
just need the number..